### PR TITLE
Add Volumes section to the service form

### PR DESCRIPTION
**Pull request by Poltergeist**
This adds the support for the volumes tab. 

image

- [x] Fix Styling (Deferred to another PR)

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @orlandohohmeier

File: src/js/schemas/service-schema/Volumes.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+const Volumes = {
+  title: 'Volumes',
+  type: 'object',
+  properties: {
+    localVolumes: {
+      type: 'array',
+      duplicable: true,
+      addLabel: 'Add LocalVolume',
```
Comment (aldipower) [inline src/js/schemas/service-schema/Volumes.js]: Missing space?
'Add Local Volume'

File: src/js/utils/ServiceUtil.js [hunk 2]
```diff
@@ -103,6 +105,56 @@ const ServiceUtil = {
...
+        let type = VolumeConstants.type.MESOS;
+        let volumesList = [];
+
+        if (definition.container == null) {
+          definition.container = {};
+        }
+
+        if (definition.container.docker &&
+          definition.container.docker.image) {
+          type = VolumeConstants.type.DOCKER;
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: A line above and below this one could increase the readability.

File: src/js/utils/ServiceUtil.js [hunk 3]
```diff
@@ -103,6 +105,56 @@ const ServiceUtil = {
...
+
+        if (volumes.localVolumes) {
+          let localVolumes = volumes.localVolumes
+            .map(function ({containerPath, size}) {
+              return {
+                containerPath,
+                persistent: {size},
+                mode: VolumeConstants.mode.rw
+              };
+            });
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: Extra line below this one?
Little bit too dense here in my opinion.

File: src/js/structs/Service.js [hunk 4]
```diff
@@ -120,6 +120,10 @@ module.exports = class Service extends Item {
     };
   }
 
+  getResidency() {
```
Comment (aldipower) [inline src/js/structs/Service.js]: Could you add a simple test?
The other props have one too.

Same for getUpdateStrategy().

**Comment by aldipower** [comment]
Besides some code readability issues and the lack of tests, this looks good to me!
Have to test it in the browser though..

File: src/js/structs/Service.js (hunk 4 above)
Comment (Poltergeist) [inline src/js/structs/Service.js]: added tests.

File: src/js/utils/ServiceUtil.js [hunk 5]
```diff
@@ -103,6 +105,59 @@ const ServiceUtil = {
         }
       }
 
+      if (volumes != null) {
```
Comment (aldipower) [inline src/js/utils/ServiceUtil.js]: Testing this would be nice and important too. :)

File: src/js/schemas/service-schema/Volumes.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+          hostPath: {
+            title: 'Host Path',
+            type: 'string'
+          },
+          mode: {
+            title: 'Mode',
+            fieldType: 'select',
+            default: 'ro',
+            options: [
+              {html: 'ReadOnly', id: 'ro'},
```
Comment (kennyt) [inline src/js/schemas/service-schema/Volumes.js]: should there be a space between `Read` and `Only`?

**Comment by kennyt** [comment]
if i create a container volume and switch it to json mode and then switch back, it doesn't save it

**Comment by Poltergeist** [comment]
@kennyt That's right to configure a docker Volume you need to set a docker image as well. So if you switch back and forth to the JSON editor you will lose this setting if you did not set a container image.

``JSON

{
  "id": "",
  "instances": 0,
  "container": {
    "docker": {
      "image": "nginx",
      "forcePullImage": false,
      "privileged": false
    },
    "type": "DOCKER",
    "volumes": [
      {
        "containerPath": "home",
        "hostPath": "home",
        "mode": "RO"
      }
    ]
  }
}
```
If you paste that into the JSON editor and switch back it should be set.

**Comment by Poltergeist** [comment]
**mesosphere-ci** retest this please

**Comment by kennyt** [comment]
testing this on cypress locally. will merge after pass

**Comment by Poltergeist** [comment]
image

**Comment by Poltergeist** [comment]
Thanks for merging @kennyt ! 🎆

**Pull request merged**